web/modules/custom/joinup_federation/src/JoinupFederationHashGenerator.php

```diff
@@ -0,0 +1,143 @@
+<?php
+
+declare(strict_types = 1);
+
+namespace Drupal\joinup_federation;
+
+use Drupal\sparql_entity_storage\Database\Driver\sparql\ConnectionInterface;
+use Drupal\sparql_entity_storage\Entity\Query\Sparql\SparqlArg;
+use Drupal\sparql_entity_storage\SparqlEntityStorageFieldHandlerInterface;
+use Drupal\sparql_entity_storage\SparqlEntityStorageGraphHandlerInterface;
+use EasyRdf\Sparql\Result;
+
+/**
+ * Helper class to generate a hash containing all data of an rdf entity.
+ *
+ * @package Drupal\joinup_federation
+ */
+class JoinupFederationHashGenerator {
+
+  /**
+   * The SPARQL database connection.
+   *
+   * @var \Drupal\rdf_entity\Database\Driver\sparql\ConnectionInterface
+   */
+  protected $connection;
+
+  /**
+   * The SPARQL storage field handler service.
+   *
+   * @var \Drupal\sparql_entity_storage\SparqlEntityStorageFieldHandler
+   */
+  protected $fieldHandler;
+
+  /**
+   * The graph handler service.
+   *
+   * @var \Drupal\sparql_entity_storage\SparqlEntityStorageGraphHandlerInterface
+   */
+  protected $graphHandler;
+
+  /**
+   * Constructs a JoinupFederationHashGenerator object.
+   *
+   * @param \Drupal\sparql_entity_storage\Database\Driver\sparql\ConnectionInterface $connection
+   *   The SPARQL database connection.
+   * @param \Drupal\sparql_entity_storage\SparqlEntityStorageFieldHandlerInterface $field_handler
+   *   The SPARQL storage field handler service.
+   * @param \Drupal\sparql_entity_storage\SparqlEntityStorageGraphHandlerInterface $graph_handler
+   *   The graph handler service.
+   */
+  public function __construct(ConnectionInterface $connection, SparqlEntityStorageFieldHandlerInterface $field_handler, SparqlEntityStorageGraphHandlerInterface $graph_handler) {
+    $this->connection = $connection;
+    $this->fieldHandler = $field_handler;
+    $this->graphHandler = $graph_handler;
+  }
+
+  /**
+   * Generates the representing hash for a given list of entity ids.
+   *
+   * The hash is generated by gathering all information related to the entity id
+   * ordered by the field predicate and values alphabetically so that the end
+   * result cannot vary between same versions.
+   *
+   * @param array $entity_ids
+   *   An array of entity ids to generate the hashes for.
+   *
+   * @return array
+   *   An associative array of hashed indexed by their related entity id.
+   */
+  public function generateDataHash(array $entity_ids): array {
```

Comment:
I will give it a shot and will let you know.

Reply:
From an example run, the following fields are not empty when converting the entity to an array:

Distributions : `id, rid, label, created, changed, uuid, path, graph, parent, field_ad_access_url, field_ad_description, field_ad_format, field_ad_licence`
Owner: `id, rid, uuid, path, graph, field_owner_name, field_owner_type`
Asset release: `id, rid, label, created, changed, uuid, path, graph, field_isr_contact_information, field_isr_description, field_isr_distribution, field_isr_is_version_of, field_isr_owner, field_isr_release_number`
Solution: `id, rid, label, created, changed, uuid, path, graph, field_is_contact_information, field_is_description, field_is_has_version, field_is_landing_page, field_is_language, field_is_owner, field_is_related_solutions, field_is_solution_type`

In the above examples, some ADMS fields are not present because it happens that the example file was not filled. The rest of the non-adms fields are filled with default values created along with the entity. Do you think it is worth using these values as well? that would make it a hash of the entity that is about to be created and not of the incoming data.

If we were to strip off the non ADMS fields then we either hardcode them or we will have to loop over the fields from the FieldHandler which makes it much less performant than querying for the data.

Reply:
As for the integrity of the hash, I tried without diving deeper on the array and it turns that one distribution produced a different hash once so I had to change it to a ksort similar to the one I already have.
With that, I tried it multiple times and the results remain the same so we are covered in terms of that.

Reply:
In general, I had something like
```
$return = [];
    foreach (Rdf::loadMultiple($entity_ids, ['staging']) as $entity_id => $entity) {
      $entity_as_array = $entity->toArray();
      $entity_as_array = array_filter($entity_as_array);

      foreach ($entity_as_array as &$values_array) {
        foreach ($values_array as &$values) {
          ksort($values);
        }
        ksort($values_array);
      }
      ksort($entity_as_array);


      $return[$entity_id] = md5(serialize($entity_as_array));
    }
    return $return;
```

Reply:
We've been decided that we keep the actual implementation as is much closer to "data from file".